src/linear_algebra/determinant.lean

```diff
@@ -103,4 +104,100 @@ instance : is_monoid_hom (det : matrix n n R → R) :=
 { map_one := det_one,
   map_mul := det_mul }
 
+/-- Transposing a matrix preserves the determinant. -/
+@[simp] lemma det_transpose (M : matrix n n R) : M.transpose.det = M.det :=
+begin
+apply sum_bij (λ σ _, σ⁻¹),
+{ intros σ _, apply mem_univ },
+{ intros σ _,
+  rw [sign_inv],
+  congr' 1,
+  apply prod_bij (λ i _, σ i),
+  { intros i _, apply mem_univ },
+  { intros i _, simp },
+  { intros i j _ _ h, simp at h, assumption },
+  { intros i _, use σ⁻¹ i, finish }
+},
+{ intros σ σ' _ _ h, simp at h, assumption },
+{ intros σ _, use σ⁻¹, finish }
+end
+
+/-- Permuting the columns changes the sign of the determinant. -/
+lemma det_permute (M : matrix n n R) (σ : perm n) : matrix.det (λ i, M (σ i)) = σ.sign * M.det :=
+begin
+unfold det,
+rw mul_sum,
+apply sum_bij (λ τ _, σ * τ),
+{ intros τ _, apply mem_univ },
+{ intros τ _,
+  show
+    ↑(sign τ) * finset.prod univ (λ i, M (σ.to_fun (τ.to_fun i)) i)
+    = ↑(sign σ) * (↑(sign (σ * τ)) * finset.prod univ (λ i, M (σ.to_fun (τ.to_fun i)) i)),
+  rw ←mul_assoc,
+  congr,
+  calc ε τ
+       = ↑(sign σ * sign σ * sign τ) :
+    by {conv_lhs {rw [←one_mul (sign τ), ←int.units_pow_two (sign σ)]}, norm_num}
+  ... = ε σ * ε (σ * τ) :
+    by simp only [mul_assoc, int.cast_mul, sign_mul, coe_coe, units.coe_mul] },
+{ intros τ τ' _ _, exact (mul_left_inj σ).mp },
+{ intros τ _, use σ⁻¹ * τ, use (mem_univ _), exact (mul_inv_cancel_left _ _).symm }
+end
+
+section det_zero
+/-! ### `det_zero` section
+
+  Prove that a matrix with a repeated column has determinant equal to zero.
+-/
+
+/--
+  `mod_swap i j` contains permutations up to swapping `i` and `j`.
+
+  We use this to partition permutations in the expression for the determinant,
+  such that each partitions sums up to `0`.
+-/
+def mod_swap {n : Type u} [decidable_eq n] (i j : n) : setoid (perm n) :=
```

Comment:
I tried rewording the statement, but unfortunately it seems like the original version is easier to work with.

The argument I want to use is that the determinant is a sum adding up to `0` because the term corresponding to `σ` cancels with the term corresponding to `swap i j * σ`. Thus, I want to partition the permutations so that the classes are `{σ, swap i j * σ}`, and then the sum over this class is `x + -1 * x + 0 = 0`.

Using the quotient of `group.closure {swap i j}` would then require as a first step showing that the classes are `{σ, swap i j * σ}`, which was immediate from the original `mod_swap` definition. (Is there a better way to show `quotient_group.left_rel (group.closure {swap i j}) σ τ ↔ (σ = τ ∨ σ = swap i j * τ)` than first using `group.closure {swap i j} = gpowers (swap i j) = {σ // ∃ k, σ = (swap i j)^k}`, then showing `(∃ k, σ = (swap i j)^k) ↔ (∃ 0 ≤ k < order_of (swap i j), σ = (swap i j)^k) ↔ (σ = id ∨ σ = swap i j)`?) So I don't expect the change is really worth the extra work.